File: src/core/scoring-engine.js
import { TASK_STATUS } from './task-status.js';

const difficultyWeight = {
  1: 0.8,
  2: 1.0,
  3: 1.2
};

export function computeIntegrityScore(tasks = []) {
  const completed = tasks.filter((t) => t.status === TASK_STATUS.COMPLETED);
  const missed = tasks.filter((t) => t.status === TASK_STATUS.MISSED);
  const pending = tasks.filter((t) => t.status === TASK_STATUS.PENDING);

  let rawTotal = 0;
  let maxPossible = 0;

  for (const task of tasks) {
    const impact = Number(task.estimatedImpact) || 0;
    const diffW = difficultyWeight[task.difficulty] ?? 1.0;

    if (task.status === TASK_STATUS.COMPLETED) {
      const timeW = task.onTime === false ? 0.7 : 1.0;
      const taskScore = impact * diffW * timeW;
      rawTotal += taskScore;
    } else if (task.status === TASK_STATUS.MISSED) {
      rawTotal -= impact;
    }

    maxPossible += impact * diffW;
  }

  if (maxPossible <= 0) {
    return {
      score: 0,
      completedCount: completed.length,
      missedCount: missed.length,
      pendingCount: pending.length,
      rawTotal,
      maxPossible
    };
  }

  const ratio = rawTotal / maxPossible;
  const clamped = Math.max(0, Math.min(1, ratio));
  const integrityScore = Math.round(clamped * 100);

  return {
    score: integrityScore,
    completedCount: completed.length,
    missedCount: missed.length,
    pendingCount: pending.length,
    rawTotal,
    maxPossible
  };
}

export function explainIntegrityScore(tasks = []) {
  const summary = computeIntegrityScore(tasks);
  const completedOnTime = tasks.filter(
    (t) => t.status === TASK_STATUS.COMPLETED && t.onTime === true
  ).length;
  const completedLate = tasks.filter(
    (t) => t.status === TASK_STATUS.COMPLETED && t.onTime === false
  ).length;
  const missed = tasks.filter((t) => t.status === TASK_STATUS.MISSED).length;
  const totalTasks = tasks.length;
  const completionRate = totalTasks ? summary.completedCount / totalTasks : 0;
  const onTimeRate = summary.completedCount ? completedOnTime / summary.completedCount : 0;

  return {
    score: summary.score,
    breakdown: {
      completedOnTime,
      completedLate,
      missed,
      totalTasks,
      completionRate,
      onTimeRate
    }
  };
}

File: backend/jyriko/domain/scoring_engine.py
"""Integrity scoring engine — port of src/core/scoring-engine.js."""
from typing import Any

from jyriko.constants import DIFFICULTY_WEIGHTS, LATE_COMPLETION_PENALTY
from jyriko.domain.task_status import TASK_STATUS_COMPLETED, TASK_STATUS_MISSED, TASK_STATUS_PENDING


def compute_integrity_score(tasks: list[dict[str, Any]] | None = None) -> dict[str, Any]:
    tasks = tasks or []
    completed = [t for t in tasks if t.get("status") == TASK_STATUS_COMPLETED]
    missed = [t for t in tasks if t.get("status") == TASK_STATUS_MISSED]
    pending = [t for t in tasks if t.get("status") == TASK_STATUS_PENDING]

    raw_total = 0.0
    max_possible = 0.0

    for task in tasks:
        impact = float(task.get("estimatedImpact") or 0)
        diff_w = DIFFICULTY_WEIGHTS.get(task.get("difficulty"), 1.0)
        if task.get("status") == TASK_STATUS_COMPLETED:
            time_w = LATE_COMPLETION_PENALTY if task.get("onTime") is False else 1.0
            raw_total += impact * diff_w * time_w
        elif task.get("status") == TASK_STATUS_MISSED:
            raw_total -= impact
        max_possible += impact * diff_w

    base = {
        "completedCount": len(completed),
        "missedCount": len(missed),
        "pendingCount": len(pending),
        "rawTotal": raw_total,
        "maxPossible": max_possible,
    }
    if max_possible <= 0:
        return {"score": 0, **base}
    clamped = max(0.0, min(1.0, raw_total / max_possible))
    return {"score": round(clamped * 100), **base}


def explain_integrity_score(tasks: list[dict[str, Any]] | None = None) -> dict[str, Any]:
    tasks = tasks or []
    summary = compute_integrity_score(tasks)
    completed_on_time = sum(
        1 for t in tasks if t.get("status") == TASK_STATUS_COMPLETED and t.get("onTime") is True
    )
    completed_late = sum(
        1 for t in tasks if t.get("status") == TASK_STATUS_COMPLETED and t.get("onTime") is False
    )
    missed_count = sum(1 for t in tasks if t.get("status") == TASK_STATUS_MISSED)
    total = len(tasks)
    completion_rate = summary["completedCount"] / total if total else 0.0
    on_time_rate = completed_on_time / summary["completedCount"] if summary["completedCount"] else 0.0
    return {
        "score": summary["score"],
        "breakdown": {
            "completedOnTime": completed_on_time,
            "completedLate": completed_late,
            "missed": missed_count,
            "totalTasks": total,
            "completionRate": completion_rate,
            "onTimeRate": on_time_rate,
        },
    }
